enhancements/ingress/aws-lb-eip-selection.md

```diff
@@ -0,0 +1,443 @@
+### Goals
+- Users are able to use EIPs for a default NLB `IngressController` at install time.
+- Check for unassociated EIPs before passing to CCM.
+
+### Non-Goals
+- Creation of EIPs in AWS.
+- Static IP usage with NLBs for OpenShift API server, DNS, Nat Gateways, LBs, Instances.
+- To assign IPs from a Customer Owned IP (CoIP) Pool when using Outposts.
+
+## Proposal
+This enhancement adds API fields in the installer and the IngressController specification
+to set the AWS EIP for the default NLB `IngressController`. The cluster ingress operator will then create
+the default `IngressController` and get the EIP assigned to the service LoadBalancer with the annotation
+`service.beta.kubernetes.io/aws-load-balancer-eip-allocations`
+Note: `service.beta.kubernetes.io/aws-load-balancer-eip-allocations` is a standard [Kubernetes service annotation](https://kubernetes.io/docs/reference/labels-annotations-taints/#service-beta-kubernetes-io-aws-load-balancer-eip-allocations)
+
+### API Extensions
+
+#### Installer Updates
+- The first API extension for setting `eipAllocations` is in the installer [Platform](https://github.com/openshift/installer/blob/master/pkg/types/aws/platform.go) type, where the new field `NetworkLoadBalancerParameters` is added as an optional field.
+
+```go
+// Platform stores all the global configuration that all machinesets
+// use.
+type Platform struct {
+...
+
+    // eipAllocations holds eipAllocations for an default AWS
+    // NLB IngressController.
+    //
+    // +optional
+    EIPAllocations *EIPAllocations `json:"eipAllocations,omitempty"`
+}
+
+// EIPAllocations holds configuration parameters for an
+// default AWS NLB IngressController. For Example: Setting AWS EIPs https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/elastic-ip-addresses-eip.html
+type EIPAllocations struct {
+    // ingressNetworkLoadBalancer is a list of IDs for Elastic IP (EIP) addresses that
+    // are assigned to the default AWS NLB IngressController.
+    // The following restrictions apply:
+    //
+    // eipAllocations can only be used with external scope, not internal.
+    // An EIP can be allocated to only a single IngressController.
+    // The number of EIP allocations must match the number of subnets that are used for the load balancer.
+	// Each EIP allocation must be unique.
+    // A maximum of 10 EIP allocations are permitted.
+    //
+    // See https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/elastic-ip-addresses-eip.html for general
+    // information about configuration, characteristics, and limitations of Elastic IP addresses.
+    //
+    // +openshift:enable:FeatureGate=InstallEIPForDefaultNLBIngressController
+    // +optional
+    // +listType=atomic
+    // +kubebuilder:validation:XValidation:rule=`self.all(x, self.exists_one(y, x == y))`,message="eipAllocations cannot contain duplicates"
+    // +kubebuilder:validation:MaxItems=10
+    IngressNetworkLoadBalancer []EIPAllocation `json:"ingressNetworkLoadBalancer"`
```

Comment:
Could the existing `PublicIpv4Pool` field be used instead to satisfy the requirements?

https://github.com/openshift/installer/blob/694d083d3332bd5b892f0098a91be6c206b18fce/pkg/types/aws/platform.go#L119-L122

I think the answer is no, IIUC users want to setup firewall rules in advance, so they need to specify explicit IP addresses. A pool won't suffice. It may be worth drawing a distinctoin between your proposal and this existing field. Would be interested in @mtulio's take on this.

Reply:
@patrickdillon is correct: No. Those are different use cases. The PublicIpv4Pool will provide a pool advertised to AWS (Public IP pool owned by the User) and the installer (CAPA) will allocate any IPs (EIPs) from that pool. The EIP proposal is requiring pre-allocated EIPs and presenting to the installer, where those can be consumed from any existing option: amazon-provided, user-pool (BYO Public IPv4), or customer-owner IP (CoIP/Outposts).